wsmaster/che-core-api-infraproxy/src/main/java/org/eclipse/che/api/infraproxy/server/InfrastructureApiService.java

```diff
@@ -0,0 +1,143 @@
+/*
+ * Copyright (c) 2012-2018 Red Hat, Inc.
+ * This program and the accompanying materials are made
+ * available under the terms of the Eclipse Public License 2.0
+ * which is available at https://www.eclipse.org/legal/epl-2.0/
+ *
+ * SPDX-License-Identifier: EPL-2.0
+ *
+ * Contributors:
+ *   Red Hat, Inc. - initial API and implementation
+ */
+package org.eclipse.che.api.infraproxy.server;
+
+import com.fasterxml.jackson.databind.JsonNode;
+import com.fasterxml.jackson.databind.ObjectMapper;
+import com.google.common.annotations.Beta;
+import com.google.common.annotations.VisibleForTesting;
+import io.swagger.annotations.Api;
+import java.io.IOException;
+import java.io.Reader;
+import java.net.URI;
+import javax.inject.Inject;
+import javax.inject.Named;
+import javax.ws.rs.DELETE;
+import javax.ws.rs.GET;
+import javax.ws.rs.HEAD;
+import javax.ws.rs.OPTIONS;
+import javax.ws.rs.POST;
+import javax.ws.rs.PUT;
+import javax.ws.rs.Path;
+import javax.ws.rs.core.Response;
+import org.eclipse.che.api.core.ApiException;
+import org.eclipse.che.api.core.ForbiddenException;
+import org.eclipse.che.api.core.rest.Service;
+import org.eclipse.che.api.workspace.server.spi.InfrastructureException;
+import org.eclipse.che.api.workspace.server.spi.RuntimeInfrastructure;
+import org.eclipse.che.commons.annotation.Nullable;
+
+/**
+ * We use this to give our clients the direct access to the underlying infrastructure REST API. This
+ * is only allowed when we can properly impersonate the user - e.g. on OpenShift with OpenShift
+ * OAuth switched on.
+ */
+@Api(InfrastructureApiService.PATH_PREFIX)
+@Beta
+@Path(InfrastructureApiService.PATH_PREFIX)
+public class InfrastructureApiService extends Service {
+  static final String PATH_PREFIX = "/unsupported/k8s";
```

Comment:
this is our chance to have our `unsafe` and I like it maybe more than `unsupported`. WDYT?

Reply:
Not sure. The endpoint is not really unsafe, we only allow it when it is safe to do so. But IMHO we do want stress, this is a temporary measure. I thought about `experimental` - but that could stay for eternity without raising eyebrows. `unsupported` on the other hand would have those eyebrows raised, I hope :smile: